client/files.go

```diff
@@ -0,0 +1,162 @@
+
+import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+	"io"
+	"mime/multipart"
+	"net/textproto"
+	"strings"
+)
+
+// PushOptions contains the options for a call to Push.
+type PushOptions struct {
+	// Source is the source of data to write (required).
+	Source io.Reader
+
+	// Path indicates the absolute path of the file in the destination
+	// machine (required).
+	Path string
+
+	// MakeDirs, if true, will create any non-existing directories in the path
+	// to the remote file. If false, the default, the call to Push will
+	// fail if any non-existing directory is found on the remote path.
+	MakeDirs bool
+
+	// Permissions indicates the mode of the file in the destination machine.
+	// Defaults to 0644. Note that, when used together with MakeDirs, the
+	// directories that might be created will not use this mode, but 0755.
+	Permissions string
+
+	// UserID indicates the user ID of the owner for the file in the destination
+	// machine. When used together with MakeDirs, the directories that might be
+	// created will also be owned by this user.
+	UserID *int
+
+	// User indicates the name of the owner user for the file in the destination
+	// machine. When used together with MakeDirs, the directories that might be
+	// created will also be owned by this user.
+	User string
+
+	// GroupID indicates the ID of the owner group for the file in the
+	// destination machine. When used together with MakeDirs, the directories
+	// that might be created will also be owned by this group.
+	GroupID *int
+
+	// Group indicates the name of the owner group for the file in the
+	// destination machine. When used together with MakeDirs, the directories
+	// that might be created will also be owned by this group.
+	Group string
+}
+
+type writeFilesPayload struct {
+	Action string           `json:"action"`
+	Files  []writeFilesItem `json:"files"`
+}
+
+type writeFilesItem struct {
+	Path        string `json:"path"`
+	MakeDirs    bool   `json:"make-dirs"`
+	Permissions string `json:"permissions"`
+	UserID      *int   `json:"user-id"`
+	User        string `json:"user"`
+	GroupID     *int   `json:"group-id"`
+	Group       string `json:"group"`
+}
+
+type fileResult struct {
+	Path  string `json:"path"`
+	Error *Error `json:"error,omitempty"`
+}
+
+var quoteEscaper = strings.NewReplacer("\\", "\\\\", `"`, "\\\"")
+
+func escapeQuotes(s string) string {
+	return quoteEscaper.Replace(s)
+}
+
+// Push writes content to a path on the remote system.
+func (client *Client) Push(opts *PushOptions) error {
+	// Buffer for multipart header/footer
+	var b bytes.Buffer
+	mw := multipart.NewWriter(&b)
+
+	// Encode metadata part of the header
+	part, err := mw.CreatePart(textproto.MIMEHeader{
+		"Content-Type":        {"application/json"},
+		"Content-Disposition": {`form-data; name="request"`},
+	})
+	if err != nil {
+		return fmt.Errorf("cannot encode metadata in request payload: %w", err)
+	}
+
+	payload := writeFilesPayload{
+		Action: "write",
+		Files: []writeFilesItem{{
```

Comment:
I see the daemon (and the slices [] of writeFileItem confirms) can handle a body with multiple file parts. Any idea how that would or could map to the push command we expose? Perhaps in the future we could support transfer multiple files by specifying a directory, or recursively write files from a base directory?

I guess this is not important, just noticed the difference between the daemon and client.

Reply:
Interesting feedback. I limited myself to implement the API that was already on the spec, but I assume we could take a `[]*PushOptions` slice instead of just `*PushOptions` in order to support this without much more complexity.

Reply:
Yeah, we spent time designed the underlying API to be multi-file friendly, but we decided to only implement single-file Go functions in the Client (Go client, as well as the Python client in `ops`) for simplicity for now. We can add them separately later if need be.

Reply:
Sounds like this can be a future discussion.